design-documents/service-communication-exception-handling.md

```diff
@@ -0,0 +1,107 @@
+
+#### Examples
+
+Twitter
+
+```
+{
+    "errors": [
+        {
+            "code": 215,
+            "message": "Bad Authentication data."
+        }
+    ]
+}
+```
+
+Facebook
+
+```
+{
+    "error": {
+        "message": "An active access token must be used to query information about the current user.",
+        "type": "OAuthException",
+        "code": 2500,
+        "fbtrace_id": "DzkTMkgIA7V"
+    }
+}
+```
+
+Twilio
+
+```
+<?xml version='1.0' encoding='UTF-8'?>
+<TwilioResponse>
+    <RestException>
+        <Code>20404</Code>
+        <Message>The requested resource /2010-04-01/Accounts/1234/IncomingPhoneNumbers/1234 was not found</Message>
+        <MoreInfo>https://www.twilio.com/docs/errors/20404</MoreInfo>
+        <Status>404</Status>
+    </RestException>
+</TwilioResponse>
+```
+
+### Exception class name or class name
+
+Existing REST API returns only error message, so we need to add exception class name or error code in addition to HTTP response code. This would allow different consumers of the service handle errors or understand what exception need to be handled or which exception type need to be thrown.
+
+Return exception class names
+Pros
+    * No mapping needed, exceptions from API can easily be caught or thrown
+Cons
+    * If consumer or producer service has non PHP or Magento implementation PHP exception class name doesn't make a lot of sense
+    
+Return error code
+Pros
+    * Error codes make sense regardless of technology
+Cons
+    * We would have to maintain some sort of mapping to make it work.
+    
+Replaceability of services is one of the goals of the project, so the second option makes more sense.
+
+There are could be third option, which is combination of both options (what Facebook does), it also makes sense but we have some data duplication in this case.
+
+#### Error code as HTTP status code or in message
+
+Some APIs put status code in the HTTP response code (introduce custom codes). I don't have strong openion on this, so I went with approach that Facebook, Twitter and Twilio use.
+
+### Throw or suppress exceptions in consumers
+
+All exceptions except network exceptions we should map to exceptions declared on the service contracts that communicate through the network.
+
+When we consume services in BFF we could benefit from knowing that network exception happen, because we could make UI more friendly. For instance, we were not able to calculate totals on shopping cart page. We can still show shopping art and display the message that we retrying total calculation.
+
+In order to make this improvements in UI, we would have to declare new type of exception on service contracts `DataRetrievalFailedException`. This is would be backwards incompatible change for monolith.
```

Comment:
Does this exception work only for GET operations? What about POST, PUT, DELETE? I think, for the service, the exception might be the same as now `NotFoundException`, `CouldNotSaveException`, etc. Because the service doesn't know about the network communication. In general, the service might be called via REST, a message queue, etc.

Reply:
We can't replace `DataRetrievalFailedException` with `NotFoundException` because you may need to handle them differently. I will think about exceptions for POST, PUT and DELETE methods.